# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [29]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced
from sklearn.metrics import classification_report

import warnings
warnings.filterwarnings('ignore')

# Additional Modules
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from imblearn.over_sampling import RandomOverSampler


---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [9]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending = pd.read_csv("Resources/lending_data.csv")
# Review the DataFrame
## Review the first 5 rows of the DataFrame
display(df_lending.head())

# Review the last 5 rows of the DataFrame
display(df_lending.tail())

# Get the summary statistics of the DataFrame
display(df_lending.describe())

# Get the number of rows and columns in the DataFrame
display(df_lending.shape)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt   loan_status  
count     77536.000000      77536.000000  77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804      0.032243  
std           1.904426          0.582086   8371.635077      0.176646  
min           0.000000          0.000000      0.000000      0.000000  
25%           3.000000          0.000000  14800.000000      0.000000  
50%           4.000000          0.000000  18100.000000      0.000000  
75%           4.000000          1.000000  21400.000000      0.000000  
max          16.000000          3.000000  75200.000000      1.000000

(77536, 8)

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [8]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending['loan_status']

# Separate the X variable, the features
X = df_lending.drop("loan_status", axis=1)

In [15]:
# Review the y variable Series
# Display the first 5 rows of the y Series
display(y.head())

# Display the shape of the y Series
display(y.shape)

# Display the value counts of the y Series
display(y.value_counts())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

(77536,)

0    75036
1     2500
Name: loan_status, dtype: int64

In [14]:
# Review the X variable DataFrame
# Display the first 5 rows of the X DataFrame
display(X.head())

# Display the shape of the X DataFrame
display(X.shape)

# Display the column names of the X DataFrame
display(X.columns)


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

(77536, 7)

Index(['loan_size', 'interest_rate', 'borrower_income', 'debt_to_income',
       'num_of_accounts', 'derogatory_marks', 'total_debt'],
      dtype='object')

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [12]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [16]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

# Display the shapes of the training and testing datasets
display("X_train shape:", X_train.shape)
display("X_test shape:", X_test.shape)
display("y_train shape:", y_train.shape)
display("y_test shape:", y_test.shape)

'X_train shape:'

(62028, 7)

'X_test shape:'

(15508, 7)

'y_train shape:'

(62028,)

'y_test shape:'

(15508,)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [21]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
lr_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model.fit(X_train, y_train)


LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [22]:
# Make a prediction using the testing data
lr_predictions=lr_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [25]:
# Print the balanced_accuracy score of the model
balanced_acc = balanced_accuracy_score(y_test, lr_predictions)
display("Balanced accuracy score:", balanced_acc)

'Balanced accuracy score:'

0.9521352751368186

In [27]:
# Generate a confusion matrix for the model
cm = confusion_matrix(y_test, lr_predictions)
print("Confusion matrix:\n", cm)

Confusion matrix:
 [[14926    75]
 [   46   461]]


In [30]:
# Print the classification report for the model
report = classification_report(y_test, lr_predictions)
print("Classification report:\n", report)

Classification report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     15001
           1       0.86      0.91      0.88       507

    accuracy                           0.99     15508
   macro avg       0.93      0.95      0.94     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

In [32]:
cm_imbalanced_df = pd.DataFrame(cm, 
                                index = ['Actual Healthy Loans (low-risk)', 
                                'Actual Non-Healthy Loans (high-risk)'], 
                                columns = ['Predicted Healthy Loans (low-risk)', 'Predicted Non-Healthy Loans (high-risk)']
                              )
cm_imbalanced_df

Predicted Healthy Loans (low-risk)  \
Actual Healthy Loans (low-risk)                                    14926   
Actual Non-Healthy Loans (high-risk)                                  46   

                                      Predicted Non-Healthy Loans (high-risk)  
Actual Healthy Loans (low-risk)                                            75  
Actual Non-Healthy Loans (high-risk)                                      461

**Answer:**

The logistic regression model achieved a high balanced accuracy score of 95%, which indicates that the model performs well in terms of overall accuracy. However, it's important to note that the dataset is imbalanced, with a large majority of healthy loans (class 0) compared to high-risk loans (class 1). This imbalance could impact the model's ability to predict high-risk loans accurately.

Looking at the classification report, the model achieved a 100% precision score for predicting healthy loans (class 0), which is promising. However, the model achieved a lower precision score of 85% for predicting high-risk loans (class 1). This indicates that the model is more likely to incorrectly predict high-risk loans as healthy loans, which could be a cause for concern.

Overall, while the model's high balanced accuracy score is promising, the imbalanced nature of the dataset and the lower precision score for high-risk loans suggest that the model may not be optimal for predicting high-risk loans accurately.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [34]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros_model.fit_resample(X_train, y_train)

In [35]:
# Count the distinct values of the resampled labels data
display(y_resampled.value_counts())

0    60035
1    60035
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [36]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
lr_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
lr_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
lr_prediction_resampled = lr_model.predict(X_test)

# Print the first 10 predictions
display(lr_prediction_resampled[:10])


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [47]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_resampled = balanced_accuracy_score(y_test, lr_prediction_resampled)
print("Balanced Accuracy Score (resampled data):", balanced_accuracy_resampled)

Balanced Accuracy Score (resampled data): 0.9941749445500477


In [43]:
# Generate a confusion matrix for the model
cf_matrix_resampled = confusion_matrix(y_test, lr_prediction_resampled)
display("Confusion Matrix (resampled data):\n", cf_matrix_resampled)

'Confusion Matrix (resampled data):\n'

array([[14915,    86],
       [    3,   504]], dtype=int64)

In [46]:
# Print the classification report for the model
cls_report_resampled = classification_report(y_test, lr_prediction_resampled)
print("Classification Report (resampled data):\n", cls_report_resampled)

Classification Report (resampled data):
               precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.85      0.99      0.92       507

    accuracy                           0.99     15508
   macro avg       0.93      0.99      0.96     15508
weighted avg       1.00      0.99      0.99     15508



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?



In [48]:
cf_matrix_resampled_df = pd.DataFrame(cf_matrix_resampled, 
                                index = ['Actual Healthy Loans (low-risk)', 
                                'Actual Non-Healthy Loans (high-risk)'], 
                                columns = ['Predicted Healthy Loans (low-risk)', 'Predicted Non-Healthy Loans (high-risk)']
                              )
cf_matrix_resampled_df

Predicted Healthy Loans (low-risk)  \
Actual Healthy Loans (low-risk)                                    14915   
Actual Non-Healthy Loans (high-risk)                                   3   

                                      Predicted Non-Healthy Loans (high-risk)  
Actual Healthy Loans (low-risk)                                            86  
Actual Non-Healthy Loans (high-risk)                                      504

**Answer:** 
Based on the classification report for the model fit with oversampled data, we can see that the recall for the high-risk loans (label 1) is significantly higher at 0.99 compared to the recall of 0.85 for the same label in the model fit with original data. This indicates that the model fit with oversampled data is better at correctly identifying high-risk loans, which is important for a lending company.

A lending company might want a model that requires a higher recall for high-risk loans because misclassifying high-risk loans as healthy loans (label 0) can be costly for the lending company, since it may lead to a loss of funds provided by the lender. On the other hand, misclassifying healthy loans as high-risk loans is not as big of a deal since no funds have been provided to the customer, so the loss is limited to only the cost of identifying and investigating the loan further. Therefore, a model with a higher recall for high-risk loans can help the lending company avoid potential losses and manage risk more effectively.